crates/functional/src/lib.rs

```diff
@@ -181,6 +182,194 @@ where
     );
 }
 
+pub fn use_effect<F, Destructor>(callback: F)
+where
+    F: FnOnce() -> Destructor + 'static,
+    Destructor: FnOnce() + 'static,
+{
+    let callback = Box::new(callback);
+    struct UseEffectState<Destructor> {
+        destructor: Option<Box<Destructor>>,
+    }
+    use_hook(
+        |_: &mut UseEffectState<Destructor>, hook_update| {
+            return move || {
+                hook_update(move |state: &mut UseEffectState<Destructor>| {
+                    if let Some(de) = state.destructor.take() {
+                        de();
+                    }
+                    let new_destructor = callback();
+                    state.destructor.replace(Box::new(new_destructor));
+                    false
+                });
+            };
+        },
+        || UseEffectState { destructor: None },
+    )();
+}
+
+pub fn use_effect1<F, Destructor, T1>(callback: F, o1: T1)
+where
+    F: FnOnce(&T1) -> Destructor + 'static,
+    Destructor: FnOnce() + 'static,
+    T1: PartialEq + 'static,
+{
+    let callback = Box::new(callback);
+    use_effect5(
+        Box::new(|a: &T1, _: &(), _: &(), _: &(), _: &()| callback(a)),
+        o1,
+        (),
+        (),
+        (),
+        (),
+    );
+}
+
+pub fn use_effect2<F, Destructor, T1, T2>(callback: F, o1: T1, o2: T2)
+where
+    F: FnOnce(&T1, &T2) -> Destructor + 'static,
+    Destructor: FnOnce() + 'static,
+    T1: PartialEq + 'static,
+    T2: PartialEq + 'static,
+{
+    let callback = Box::new(callback);
+    use_effect5(
+        Box::new(|a: &T1, b: &T2, _: &(), _: &(), _: &()| callback(a, b)),
+        o1,
+        o2,
+        (),
+        (),
+        (),
+    );
+}
+
+pub fn use_effect3<F, Destructor, T1, T2, T3>(callback: F, o1: T1, o2: T2, o3: T3)
+where
+    F: FnOnce(&T1, &T2, &T3) -> Destructor + 'static,
+    Destructor: FnOnce() + 'static,
+    T1: PartialEq + 'static,
+    T2: PartialEq + 'static,
+    T3: PartialEq + 'static,
+{
+    let callback = Box::new(callback);
+    use_effect5(
+        Box::new(|a: &T1, b: &T2, c: &T3, _: &(), _: &()| callback(a, b, c)),
+        o1,
+        o2,
+        o3,
+        (),
+        (),
+    );
+}
+
+pub fn use_effect4<F, Destructor, T1, T2, T3, T4>(callback: F, o1: T1, o2: T2, o3: T3, o4: T4)
+where
+    F: FnOnce(&T1, &T2, &T3, &T4) -> Destructor + 'static,
+    Destructor: FnOnce() + 'static,
+    T1: PartialEq + 'static,
+    T2: PartialEq + 'static,
+    T3: PartialEq + 'static,
+    T4: PartialEq + 'static,
+{
+    let callback = Box::new(callback);
+    use_effect5(
+        Box::new(|a: &T1, b: &T2, c: &T3, d: &T4, _: &()| callback(a, b, c, d)),
+        o1,
+        o2,
+        o3,
+        o4,
+        (),
+    );
+}
+
+pub fn use_effect5<F, Destructor, T1, T2, T3, T4, T5>(
```

Comment:
I think we could make this a bit more user-friendly. How about something like

```rust
pub fn use_effect_with_deps<F, Destructor, Dependents>(
    callback: F,
    deps: Dependents
) where
    F: FnOnce(&D) -> Destructor + 'static,
    Destructor: FnOnce() + 'static,
    Dependents: PartialEq + 'static,
{
    struct UseEffectState<D, Destructor> {
        deps: Rc<D>,
        destructor: Option<Box<Destructor>>,
    }
```

Then, users can specify tuples if they need to pass different types...

```rust
use_effect_with_deps(
  move |(_num, _string)| {
    || {}
  },
  (0, String::from("hello"))
);
```

So no need for methods for each number of arguments

Reply:
Good point